bundle/v2/client.go

```diff
@@ -0,0 +1,585 @@
+// Copyright 2021-2024 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package v2
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"time"
+	"unicode"
+
+	"connectrpc.com/connect"
+	"github.com/go-logr/logr"
+	"github.com/rogpeppe/go-internal/cache"
+	"github.com/sourcegraph/conc/pool"
+	"go.uber.org/multierr"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/cerbos/cloud-api/base"
+	"github.com/cerbos/cloud-api/bundle"
+	"github.com/cerbos/cloud-api/bundle/clientcache"
+	bootstrapv2 "github.com/cerbos/cloud-api/genpb/cerbos/cloud/bootstrap/v2"
+	bundlev2 "github.com/cerbos/cloud-api/genpb/cerbos/cloud/bundle/v2"
+	"github.com/cerbos/cloud-api/genpb/cerbos/cloud/bundle/v2/bundlev2connect"
+)
+
+const bootstrapV2PathPrefix = "bootstrap/v2"
+
+type Client struct {
+	rpcClient bundlev2connect.CerbosBundleServiceClient
+	cache     *clientcache.ClientCache
+	base.Client
+}
+
+func NewClient(conf bundle.ClientConf, baseClient base.Client, options []connect.ClientOption) (*Client, error) {
+	if err := conf.Validate(); err != nil {
+		return nil, err
+	}
+
+	c, err := clientcache.New(conf.CacheDir, conf.TempDir)
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient := baseClient.StdHTTPClient() // Bidi streams don't work with retryable HTTP client.
+	return &Client{
+		Client:    baseClient,
+		rpcClient: bundlev2connect.NewCerbosBundleServiceClient(httpClient, baseClient.APIEndpoint, options...),
+		cache:     c,
+	}, nil
+}
+
+type hashStringFn func(value string) string
+
+func HashedBootstrapConfName(hashStringFn hashStringFn, clientID, clientSecret, bundleLabel string) string {
```

Comment:
I don't think the hash function should be parameterised. `Credentials.HashString` includes the workspace secret key in the hash, which is not correct, because in v2 we won't have a workspace secret.

I think the `Credentials` struct should have a `HashedClientSecret` field that is computed on creation (since we use that in multiple places, we don't need to recompute it every time).